Update developer guide (#180)
AY2223S1-CS2103T-T10-1/tp: docs/DeveloperGuide.md

#### Implementation

Goto module mechanism is facilitated by the `GoToCommand` and `GoToCommandParser`.

It allows users to navigate to a specific module given their respective module code, displaying information
(i.e. tasks, links and contacts) that are associated to that module.<br>

It uses the following methods provided by `ModelManager` which implements the `Model` interface.

> * `ModelManager::getModuleUsingModuleCode`: Retrieves a `Module` object using the `ModuleCode` object associated wih that `Module`

Reviewer: ```suggestion
* `ModelManager#getModuleUsingModuleCode()`: Retrieves a `Module` object using the `ModuleCode` object associated with that `Module`
```
See PR #185

> * `ModelManager::updateFilteredModuleList`: Update the current module list and filter it according to the given predicate `Predicate<Module> predicate`, reflecting the changes accordingly in the GUI

Reviewer: ```suggestion
* `ModelManager#updateFilteredModuleList()`: Update the current module list and filter it according to the given predicate `Predicate<Module> predicate`, reflecting the changes accordingly in the GUI
```

> * `ModelManager::updateFilteredPersonList`: Update the current person list and filter it according to the given predicate `Predicate<Person> predicate`, reflecting the changes accordingly in the GUI

Reviewer: ```suggestion
* `ModelManager#updateFilteredPersonList()`: Update the current person list and filter it according to the given predicate `Predicate<Person> predicate`, reflecting the changes accordingly in the GUI
```

> * `ModelManager::setHomeStatus`: Sets the home status of Plannit.

Reviewer: ```suggestion
* `ModelManager#setHomeStatus()`: Sets the home status of Plannit.
```

goto CS1231
```

**Step 2**: The `LogicManager` uses the `AddressBookParser` and `GoToCommandParser`
to parse the user input. After validating the arguments provided by the user, the user input
is used to perform the following actions: <br>

> * Extract the `ModuleCode` of the module to navigate to in the `GoToCommandParser` using the method `ParserUtil::parseModuleCode`

Reviewer: ```suggestion
* Extract the `ModuleCode` of the module to navigate to in the `GoToCommandParser` using the method `ParserUtil#parseModuleCode()`
```

* Instantiate a `ModuleCodeMatchesKeywordPredicate` object

**Step 3**: A `GoToCommand` object is instantiated using the `ModuleCodeMatchesKeywordPredicate` and `ModuleCode`
obtained in **Step 2** which is returned to the `LogicManager`.

> **Step 4**: `LogicManager` calls the `GoToCommand::execute` method. This method will first

Reviewer: ```suggestion
**Step 4**: `LogicManager` calls the `GoToCommand#execute()` method. This method will first
```

obtain the `Module` associated with the `ModuleCode` obtained in **Step 2** by calling

> `ModelManager::getModuleUsingModuleCode`.

Reviewer: ```suggestion
`ModelManager#getModuleUsingModuleCode()`.
```

> **Step 5**: The module list is then filtered using the `ModelManager::updateFilteredModuleList` method according

Reviewer: ```suggestion
**Step 5**: The module list is then filtered using the `ModelManager#updateFilteredModuleList()` method according
```

to the `ModuleCodeMatchesKeywordPredicate` object instantiated in **Step 2**.

**Step 6**: A `PersonIsInModulePredicate` object is then instantiated with the `Module` object obtained in **Step 4**.

> **Step 7**: The person list is then filtered using the `ModelManager::updateFilteredPersonList` method according

Reviewer: ```suggestion
**Step 7**: The person list is then filtered using the `ModelManager#updateFilteredPersonList()` method according
```

to the `PersonIsInModulePredicate` object instantiated in **Step 6**.

> **Step 8**: The home status of Plannit is set to false via `ModelManager::SetHomeStatus`.

Reviewer: ```suggestion
**Step 8**: The home status of Plannit is set to false via `ModelManager#SetHomeStatus()`.
```

inputting the `find-module` command followed by a keyword.
E.g.:
```
find-module CS
```

**Step 2**: The `LogicManager` uses the `AddressBookParser` and `FindModuleCommandParser`
to parse the user input. After validating the arguments provided by the user, the user input
is used to instantiate a `ModuleCodeStartsWithKeywordPredicate` object.

**Step 3**: A `FindModuleCommand` object is instantiated using the `ModuleCodeStartsWithKeywordPredicate`
obtained in **Step 2** which is returned to the `LogicManager`.

**Step 4**: `LogicManager` calls the `FindModuleCommand#execute()` method. This method will first
obtain the home status of Plannit by calling `Model#getHomeStatusAsBoolean()` and check if Plannit
is currently at home.

**Step 5**: The module list is then filtered using the `Model#updateFilteredModuleList()` method according
to the `ModuleCodeStartsWithKeywordPredicate` object instantiated in **Step 2**.

**Step 6**: A new `CommandResult` object is returned, indicating success.

The following sequence diagram summarizes what happens when a user executes the `find-module` command:

![FindModuleSequenceDiagram](images/FindModuleSequenceDiagram.png)

> <div markdown="span" class="alert alert-info">:information_source: **Note:** The lifeline for `GoToCommandParser`

Reviewer 2: ```suggestion
<div markdown="span" class="alert alert-info">:information_source: **Note:** The lifeline for `FindModuleCommandParser` and `FindModuleCommand`
```
Wrong command?

Author: I shall amend that thank you!

* `Model#getHomeStatusAsBoolean()`: obtains the home status of Plannit.

Given below is an example usage scenario and how the mechanism
behaves when a user list all modules in Plannit.

**Step 1**. The user requests to list all modules in Plannit by
inputting the `list-module` command.
E.g.:
```
list-module CS
```

**Step 2**: The `LogicManager` uses the `AddressBookParser` to parse the user input.
After validating the arguments provided by the user, the user input is used to instantiate
a `ListModuleCommand` object, which is returned to the `LogicManger`.

**Step 3**: `LogicManager` calls the `ListModuleCommand#execute()` method. This method will first
obtain the home status of Plannit by calling `Model#getHomeStatusAsBoolean()` and check if Plannit
is currently at home.

**Step 4**: The module list is then filtered using the `Model#updateFilteredModuleList()` method according
to the `Model#PREDICATE_SHOW_ALL_MODULES` predicate.

**Step 5**: A new `CommandResult` object is returned, indicating success.

The following sequence diagram summarizes what happens when a user executes the `list-module` command:

![ListModuleSequenceDiagram](images/ListModuleSequenceDiagram.png)

> <div markdown="span" class="alert alert-info">:information_source: **Note:** The lifeline for `GoToCommandParser`

Reviewer 2: Wrong command?
```suggestion
<div markdown="span" class="alert alert-info">:information_source: **Note:** The lifeline for `ListModuleCommandParser`
```

Author: I shall amend that thank you

* Instantiate a `ModuleCodeMatchesKeywordPredicate` object

**Step 3**: A `GoToCommand` object is instantiated using the `ModuleCodeMatchesKeywordPredicate` and `ModuleCode`
obtained in **Step 2** which is returned to the `LogicManager`.

**Step 4**: `LogicManager` calls the `GoToCommand#execute()` method. This method will first
obtain the `Module` associated with the `ModuleCode` obtained in **Step 2** by calling
`Model#getModuleUsingModuleCode()`.

**Step 5**: The module list is then filtered using the `Model#updateFilteredModuleList()` method according
to the `ModuleCodeMatchesKeywordPredicate` object instantiated in **Step 2**.

**Step 6**: A `PersonIsInModulePredicate` object is then instantiated with the `Module` object obtained in **Step 4**.

**Step 7**: The person list is then filtered using the `Model#updateFilteredPersonList()` method according
to the `PersonIsInModulePredicate` object instantiated in **Step 6**.

**Step 8**: The home status of Plannit is set to false via `Model#SetHomeStatus()`.

**Step 9**: A new `CommandResult` object is returned, indicating success.

The following sequence diagram summarizes what happens when a user executes the `goto` command:

> 

Reviewer 2: For the PlantUML limitation disclaimer note below, should it also include `GoToCommand`?

Author: It have since been added thank you!

![GoToActivityDiagram](images/GoToActivityDiagram.png)

#### Navigate to home feature

##### Implementation

Navigate to home mechanism is facilitated by the `HomeCommand`.

It navigates users back to the home page of Plannit, showing all modules and persons stored in Plannit.<br>

It uses the following methods provided by the `Model` interface.
* `Model#goToHomePage()`: Sets home status to true and update module list and person list to show all modules and persons respectively, reflecting the changes accordingly in the GUI

Given below is an example usage scenario and how the mechanism
behaves when a user navigates back to the home page of Plannit.

**Step 1**. The user requests to go home by inputting the `home` command.
E.g.:
```
home
```

**Step 2**: The `LogicManager` uses the `AddressBookParser` to parse the user input.

> After validating the arguments provided by the user, the user input is used to instantiate

Reviewer 2: No need to validate arguments? `home` has no arguments and any extras will be ignored

Author: I shall amend that. Initially thought that `home` it self is also considered as an argument. thank you

a `HomeCommand` object, which is returned to the `LogicManger`.

**Step 3**: `LogicManager` calls the `HomeCommand#execute()` method. This method will then
call the `Model#goToHomePage()` method which perform the following actions:
* Set home status to true
* Update module list to show all modules
* Update person list to show all persons

**Step 5**: A new `CommandResult` object is returned, indicating success.

The following sequence diagram summarizes what happens when a user executes the `home` command:

> ![HomeSequenceDiagram](images/HomeSequenceDiagram.png)

Reviewer 2: Perhaps add PlantUML limitation disclaimer on the end-of-lifeline cross here?

Author: I shall include that

    * Cons: Require user to navigate back to home page before going to another module which might be unintuitive for users.

##### Rationale behind current choice:
1. Alternative 1 was chosen as it provides a more intuitive way for prospective users.

2. The following commands `list-module` and `find-module` are disabled as they are not meant to be used in tandem with `goto` command
   as the filtering of modules is not needed when there exist only one module.

3. The following commands `list-person` and `find-person` are disabled as they are not meant to be used in tandem with `goto` command

>    as essential person within a module (i.e. professors, tutor assistant and friends) are likely to be limited in numbers.

Reviewer 2: ```suggestion
   as essential persons within a module (i.e. professors, tutor assistant and friends) are likely to be limited in numbers.
```
Grammatical error?

Author: thanks for catching that

  Use case resumes at step 2.

#### Use case: UC03 - Find module
**Preconditions**:
* User is on the home page.

**Main Success Scenario (MSS)**
1. User chooses to search up on modules with module code starting with specific keyword.
2. Plannit requests for keyword.
3. User enters the keyword.

> 4. Plannit searches and displays a list of module with mode code starting with the specified keyword.

Reviewer 2: ```suggestion
4. Plannit searches and displays a list of modules with module code starting with the specified keyword.
```
Typo here?

Author: Thank you for catching that!

### Deleting a module

1. Deleting a module while all modules are being shown at home page

>     1. Prerequisites: User is at home page. List all modules using the `list-module` command. Module `CS1231S` exist in the list.

Reviewer: ```suggestion
    1. Prerequisites: User is at home page. List all modules using the `list-module` command. Module `CS1231S` exists in the list.
```

Author: thanks Ive since updated

    2. Test case: `delete-module m/CS1231S`<br>
       Expected: `CS1231S` is deleted from the list. Details of the deleted module shown in the status message. Timestamp in the status bar is updated.

    3. Test case: `delete-module`<br>
       Expected: No module is deleted. Error details shown in the status message. Status bar remains the same.

    4. Other incorrect delete commands to try: `delete-module m/` <br>
       Expected: Similar to previous.

1. _{ more test cases … }_

### Deleting a contact

1. Deleting a contact while all contacts are being shown at home page

>     1. Prerequisites: User is at home page. List all contacts using the `list-contact` command. Person `Alice` exist in the list.

Reviewer: ```suggestion
    1. Prerequisites: User is at home page. List all contacts using the `list-contact` command. Person `Alice` exists in the list.
```

Author: I have since updated it thank you!
